apps/idos-enclave/src/lib/enclave.js

```diff
@@ -248,12 +249,50 @@ export class Enclave {
     return decrypted
       .filter((credential) => {
         const content = JSON.parse(credential.content);
-
         return countries.includes(content.credentialSubject.residential_address_country);
       })
       .map((credential) => credential.id);
   }
 
+  async filterCredentials(credentials, privateFieldFilters) {
+    const filterCredentials = async (credentials, privateFieldFilters) => {
+      const filterByKeys = (content, keys, filterCriteria, include = true) => {
+        return every(keys, (key) => {
+          const matchedValue = get(content, key);
+          const isMatch = filterCriteria[key].includes(matchedValue);
+          return include ? isMatch : !isMatch;
+        });
+      };
+
+      const positiveMatchesKeys = Object.keys(privateFieldFilters.pick);
+      const omitKeys = Object.keys(privateFieldFilters.omit);
+
+      const decrypted = await Promise.all(
+        credentials.map(async (credential) => ({
+          ...credential,
+          content: Utf8Codec.decode(
+            await this.decrypt(
+              Base64Codec.decode(credential.content),
+              Base64Codec.decode(credential.encryption_public_key),
+            ),
+          ),
+        })),
+      );
+
+      const matches = decrypted.filter((credential) => {
+        const content = JSON.parse(credential.content);
+        return filterByKeys(content, positiveMatchesKeys, privateFieldFilters.pick, true);
+      });
+
+      return matches.filter((match) => {
+        const content = JSON.parse(match.content);
+        return filterByKeys(content, omitKeys, privateFieldFilters.omit, false);
+      });
+    };
+
+    return filterCredentials(credentials, privateFieldFilters);
+  }
```

Comment:
How about this, to remove more boilerplate?

```suggestion
  async filterCredentials(credentials, privateFieldFilters) {
    const filterByKeys = (content, filterCriteria) => {
      return every(Object.entries(filterCriteria), ([key, targetSet]) => {
        return targetSet.includes(get(content, key));
      });
    };

    const decrypted = await Promise.all(
      credentials.map(async (credential) => ({
        ...credential,
        content: Utf8Codec.decode(
          await this.decrypt(
            Base64Codec.decode(credential.content),
            Base64Codec.decode(credential.encryption_public_key),
          ),
        ),
      })),
    );

    return decrypted.filter((credential) => {
      const content = JSON.parse(credential.content);
      return filterByKeys(content, privateFieldFilters.pick) && !filterByKeys(content, privateFieldFilters.omit);
    });
  }
```